nanopass/nanopass.nim

```diff
@@ -0,0 +1,445 @@
+      if it.mvar in vars:
+        if vars[it.mvar] != it.typ:
+          error(fmt"cannot inherit '{form}'; '{it.mvar}' changed its meaning",
+                info)
+      else:
+        error(fmt"cannot inherit '{form}'; '{it.mvar}' was removed", info)
+
+  # after all that setup, the productions can finally be inherited:
+  for name, it in base.nterminals.pairs:
+    var res = it
+    # inherit the forms and patch references:
+    for f in res.forms.mitems:
+      # make sure the forms are still valid after the previous removals/renames
+      checkForm(base, vars, base.forms[f.form], info)
+      var idx = result.forms.find(base.forms[f.form])
+      if idx == -1:
+        idx = result.addForm(base.forms[f.form])
+      f.form = idx
+
+    # check the meta-vars:
+    for v in res.vars.items:
+      if v notin vars:
+        error(fmt"cannot inherit '{name}'; '{v}' (used as a production of '{name}') was removed",
+              info)
+
+    result.nterminals[name] = res
+
+  # ---- phase 2: make all additions
+  for it in add.items:
+    let (name, typ) = processTerminal(it)
+    checkName(result, vars, name, it)
+    result.terminals[name] = typ
+    vars[name] = name
+
+  proc addProd(def: var LangDef, n: NimNode, to: string) =
+    case n.kind
+    of nnkCall:
+      let (names, form) = parseForm(vars, n)
+      var idx = def.forms.find(form)
+      if idx == -1:
+        # register the form first
+        idx = def.addForm(form)
+      if def.nterminals[to].forms.findIt(it.form == idx) != -1:
+        error(fmt"production is already part of '{to}'", n)
+      def.nterminals[to].forms.add NamedForm(names: names, form: idx)
+    of nnkIdent:
+      let name = n.strVal
+      if name notin vars:
+        error(fmt"no meta-variable with name '{name}'", n)
+      def.nterminals[to].vars.add name
+    else:
+      error(fmt"unexpected syntax: {n.kind}", n)
+
+  # register the non-terminals first:
+  for it in def.items:
+    let name = it.name[0].strVal
+    if it.add.len > 0 and name notin base.nterminals:
+      # it's a new non-terminal
+      checkName(result, vars, name, it.name)
+      var nt = NonTerminal()
+      for i in 1..<it.name.len:
+        checkName(result, vars, it.name[i].strVal, it.name[i])
+        nt.mvars.add it.name[i].strVal
+        vars[it.name[i].strVal] = name
+
+      result.nterminals[name] = nt
+
+  # add the productions:
+  for it in def.items:
+    let name = it.name[0].strVal
+    if it.add.len > 0:
+      for a in it.add.items:
+        addProd(result, a, name)
+
+  # TODO: implement tag ID computation for terminals
+  # TODO: re-use tag IDs from the base language
+  computeTags(result)
+
+proc makeLanguage(body: NimNode): LangDef =
+  ## Creates a language definition from the ``defineLanguage`` DSL code.
+  body.expectMinLen 1
+  var add: seq[NimNode]
+  var def: seq[NonTerminalDef]
+
+  # second pass: process the productions
+  proc extract(n: NimNode, list: var seq[NimNode]) =
```

Comment:
I like the internal procs, I'm finding them quite a bit more readable than having them at the module level.